### PR TITLE
[WIFI-1149] profile unique name validation

### PR DESCRIPTION
**Pull request by ralphlee3**

### CODE REVIEW

**Review by mikehansen1970 — APPROVED** [review]
Approved, but you may want Dmitry to have a look

File: profile-datastore-common-test/src/main/java/com/telecominfraproject/wlan/profile/datastore/BaseProfileDatastoreTest.java [hunk 1]
```diff
@@ -291,10 +313,9 @@ public void testProfilePagination()
     
     @Test
     public void testChildProfiles() {
-    	int nextId = (int) testSequence.getAndIncrement();
 
-    	Profile profile_c1 = createProfileObject(nextId);
-    	Profile profile_c2 = createProfileObject(nextId);
+    	Profile profile_c1 = createProfileObject(getNextCustomerId());
+    	Profile profile_c2 = createProfileObject(getNextCustomerId());
```
Comment (dmitry-toptygin-connectus-ai) [inline profile-datastore-common-test/src/main/java/com/telecominfraproject/wlan/profile/datastore/BaseProfileDatastoreTest.java:318]: Nope, by changing code to use different customerId per profile you have changed the spirit of the test. These profiles are meant to be owned by the same customer (they are children of profile_p1). To do it right the createProfileObject has to be augmented with profileName (and/or type) arguments, and that should be used to make it work with the new constraint on the name/type.

File: profile-datastore-common-test/src/main/java/com/telecominfraproject/wlan/profile/datastore/BaseProfileDatastoreTest.java [hunk 2]
```diff
@@ -393,33 +413,32 @@ public void testGetProfileWithChildren(){
...
 
     @Test
-    public void testGetTopLevelProfiles(){
-    	int nextId = (int) testSequence.getAndIncrement();
+    public void testGetTopLevelProfiles() {
 
-    	Profile profile_c1 = createProfileObject(nextId);
-    	Profile profile_c2 = createProfileObject(nextId);
+    	Profile profile_c1 = createProfileObject(getNextCustomerId());
+    	Profile profile_c2 = createProfileObject(getNextCustomerId());
```
Comment (dmitry-toptygin-connectus-ai) [inline profile-datastore-common-test/src/main/java/com/telecominfraproject/wlan/profile/datastore/BaseProfileDatastoreTest.java:419]: same here, the profiles used in this test have to belong to the same customer

File: profile-datastore-inmemory/src/main/java/com/telecominfraproject/wlan/profile/datastore/inmemory/ProfileDatastoreInMemory.java [hunk 3]
```diff
@@ -46,6 +47,13 @@ public Profile create(Profile profile) {
     		throw new IllegalStateException("unsupported value in profile object");
     	}
     	
+    	for (Profile inMemProfile : idToProfileMap.values()) {
+    		if (profile.getName().equals(inMemProfile.getName()) && 
+    				profile.getProfileType().equals(inMemProfile.getProfileType())) {
+    			throw new DsDuplicateEntityException("Profile with the same name and type already exists");
```
Comment (dmitry-toptygin-connectus-ai) [inline profile-datastore-inmemory/src/main/java/com/telecominfraproject/wlan/profile/datastore/inmemory/ProfileDatastoreInMemory.java:53]: Unique combination for the profile attributes is customerId, profileType, and name. You're missing customerId in here.

File: profile-service-remote/src/test/java/com/telecominfraproject/wlan/profile/ProfileServiceRemoteTest.java [hunk 4]
```diff
@@ -289,27 +289,26 @@ public void testProfilePagination()
 
     @Test
     public void testGetProfileWithChildren(){
-    	int nextId = getNextCustomerId();
 
-    	Profile profile_c1 = createProfileObject(nextId);
-    	Profile profile_c2 = createProfileObject(nextId);
+    	Profile profile_c1 = createProfileObject(getNextCustomerId());
+    	Profile profile_c2 = createProfileObject(getNextCustomerId());
```
Comment (dmitry-toptygin-connectus-ai) [inline profile-service-remote/src/test/java/com/telecominfraproject/wlan/profile/ProfileServiceRemoteTest.java:294]: these profiles have to belong to the same customerId in this test

File: profile-service-remote/src/test/java/com/telecominfraproject/wlan/profile/ProfileServiceRemoteTest.java [hunk 5]
```diff
@@ -338,32 +337,31 @@ public void testGetProfileWithChildren(){
 
     @Test
     public void testGetTopLevelProfiles(){
-    	int nextId = getNextCustomerId();
 
-    	Profile profile_c1 = createProfileObject(nextId);
-    	Profile profile_c2 = createProfileObject(nextId);
+    	Profile profile_c1 = createProfileObject(getNextCustomerId());
+    	Profile profile_c2 = createProfileObject(getNextCustomerId());
```
Comment (dmitry-toptygin-connectus-ai) [inline profile-service-remote/src/test/java/com/telecominfraproject/wlan/profile/ProfileServiceRemoteTest.java:342]: these profiles have to belong to the same customerId in this test

File: profile-service/src/test/java/com/telecominfraproject/wlan/profile/controller/ProfileControllerTest.java [hunk 6]
```diff
@@ -76,6 +75,25 @@ public void testProfileCRUD() throws Exception {
         profileController.delete(ret.getId());
         
     }
+
```
Comment (dmitry-toptygin-connectus-ai) [inline profile-service/src/test/java/com/telecominfraproject/wlan/profile/controller/ProfileControllerTest.java:78]: I would move this test into ProfileServiceRemoteTest, and also add a test for the update() operation to make sure that name uniqueness is also enforced there

File: profile-datastore-inmemory/src/main/java/com/telecominfraproject/wlan/profile/datastore/inmemory/ProfileDatastoreInMemory.java (hunk 3 above)
Comment (dmitry-toptygin-connectus-ai) [reply profile-datastore-inmemory/src/main/java/com/telecominfraproject/wlan/profile/datastore/inmemory/ProfileDatastoreInMemory.java:53]: also, need to have the same check in the update() method in this class

**Review by dmitry-toptygin-connectus-ai — APPROVED** [review]

**Pull request merged**